flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/TestingInputsLocationsRetriever.java

```diff
@@ -0,0 +1,98 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler;
+
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+import org.apache.flink.runtime.taskmanager.LocalTaskManagerLocation;
+import org.apache.flink.runtime.taskmanager.TaskManagerLocation;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+/**
+ * A simple inputs locations retriever for testing purposes.
+ */
+class TestingInputsLocationsRetriever implements InputsLocationsRetriever {
+
+	private final Map<ExecutionVertexID, List<ExecutionVertexID>> producersByConsumer;
+
+	private final Map<ExecutionVertexID, CompletableFuture<TaskManagerLocation>> taskManagerLocationsByVertex = new HashMap<>();
+
+	TestingInputsLocationsRetriever(final Map<ExecutionVertexID, List<ExecutionVertexID>> producersByConsumer) {
+		this.producersByConsumer = new HashMap<>(producersByConsumer);
+	}
+
+	@Override
+	public Collection<Collection<ExecutionVertexID>> getConsumedResultPartitionsProducers(final ExecutionVertexID executionVertexId) {
+		final Map<JobVertexID, List<ExecutionVertexID>> executionVerticesByJobVertex =
+				producersByConsumer.getOrDefault(executionVertexId, Collections.emptyList())
+						.stream()
+						.collect(Collectors.groupingBy(ExecutionVertexID::getJobVertexId));
+
+		return new ArrayList<>(executionVerticesByJobVertex.values());
+	}
+
+	@Override
+	public Optional<CompletableFuture<TaskManagerLocation>> getTaskManagerLocation(final ExecutionVertexID executionVertexId) {
+		return Optional.ofNullable(taskManagerLocationsByVertex.get(executionVertexId));
+	}
+
+	public void markScheduled(final ExecutionVertexID executionVertexId) {
+		taskManagerLocationsByVertex.put(executionVertexId, new CompletableFuture<>());
+	}
+
+	public void assignTaskManagerLocation(final ExecutionVertexID executionVertexId) {
+		taskManagerLocationsByVertex.compute(executionVertexId, (key, future) -> {
+			if (future == null) {
+				return CompletableFuture.completedFuture(new LocalTaskManagerLocation());
+			}
+			future.complete(new LocalTaskManagerLocation());
+			return future;
+		});
+	}
+
+	static class Builder {
+
+		private final Map<ExecutionVertexID, List<ExecutionVertexID>> producersByConsumer = new HashMap<>();
+
+		public Builder connectConsumerToProducer(final ExecutionVertexID consumer, final ExecutionVertexID producer) {
+			producersByConsumer.compute(consumer, (key, producers) -> {
+				if (producers == null) {
+					producers = new ArrayList<>();
+				}
+				producers.add(producer);
+				return producers;
+			});
```

Comment:
`producersByConsumer.computeIfAbsent(consumer, (key) -> new ArrayList<>()).add(producer);` could be simple.